Fix setting state of timer units and enable GH actions

thozza opened the pull request:
Changes merged in PR#6 [1] made the tasks setting state of
`dnf-automatic-install.timer` and `dnf-automatic-reboot.timer` to be
always skipped. The reason is that `service_facts` contain status only
of `.service` units. This means that no `.timer` unit will be ever
listed in `service_facts`, even if it exists.

Change the implementation to use systemd module to check status of the
respective `.timer` units and act based on that information.

Also amend the conditions when these task are executed. Now they are
executed if in check_mode and a potentially necessary other condition is
met (such as enabled auto reboot) and regardless of the check_mode if
the `.timer` unit exists on the system.

Enable linting in GH actions.

I also wrote molecule tests in https://github.com/exploide/ansible-role-dnf-automatic/commit/de18578a70e8fe0b009819180f91852ca14e481b but [could not get it to work in GH actions so far](https://github.com/thozza/ansible-role-dnf-automatic/pull/1), although it works when run locally on Fedora using `podman`

Sorry again for the previously introduced issue 🙄

exploide commented:
Great, thank you. Just one question:

When the `dnf-automatic-reboot.timer` unit is installed for the first time, a  `daemon_reload` becomes necessary before using it. Hence the `daemon_reload: true` when setting the unit state. And now I'm wondering, isn't this also necessary when you use the systemd module for checking the unit's status? I mean, maybe `dnf_automatic_reboot_timer.status["LoadState"] == "loaded"` never becomes true without a `daemon_reload` on the first run? But I'm just guessing. If this could indeed be a problem, then maybe add the reload to the status checking tasks and remove it from the state setting tasks (because then, the daemon should already been reloaded).

> Enable linting in GH actions.

Cool, will try it out soon.

> I also wrote molecule tests in https://github.com/exploide/ansible-role-dnf-automatic/commit/de18578a70e8fe0b009819180f91852ca14e481b but https://github.com/thozza/ansible-role-dnf-automatic/pull/1, although it works when run locally on Fedora using podman

No problem. I never used molecule by myself so I would have need to read a bit about it, too. But it is sufficient for me now, if you say the role has been tested manually :yum:

thozza commented:
> Great, thank you. Just one question:
> 
> When the `dnf-automatic-reboot.timer` unit is installed for the first time, a `daemon_reload` becomes necessary before using it. Hence the `daemon_reload: true` when setting the unit state. And now I'm wondering, isn't this also necessary when you use the systemd module for checking the unit's status? I mean, maybe `dnf_automatic_reboot_timer.status["LoadState"] == "loaded"` never becomes true without a `daemon_reload` on the first run? But I'm just guessing. If this could indeed be a problem, then maybe add the reload to the status checking tasks and remove it from the state setting tasks (because then, the daemon should already been reloaded).

I agree with your assumption, although the timer unit seems to be always "loaded" for some reason even before it is enabled. I tried to get its state right after it is deployed to the host from the template and it showed up as "loaded".

Nevertheless, I think that using handler and `notify` is cleaner and more explicit. I'll rework the PR to use it.

> > Enable linting in GH actions.
> 
> Cool, will try it out soon.
> 
> > I also wrote molecule tests in [de18578](https://github.com/exploide/ansible-role-dnf-automatic/commit/de18578a70e8fe0b009819180f91852ca14e481b) but [thozza#1](https://github.com/thozza/ansible-role-dnf-automatic/pull/1), although it works when run locally on Fedora using podman
> 
> No problem. I never used molecule by myself so I would have need to read a bit about it, too. But it is sufficient for me now, if you say the role has been tested manually yum

I managed to get it working :partying_face:  I'll push it to this PR.

thozza commented:
I updated the PR. The GitHub actions can be seen run on my fork - https://github.com/thozza/ansible-role-dnf-automatic/pull/2

exploide commented:
Thanks a lot! This made me reading about GitHub Actions and molecule.
Now I got two new questions :sweat_smile: 

- We are three levels deep into containers here. The Ubuntu runner (which is actually a container I guess), the Fedora Docker container which runs molecule and the Podman container with the system under test. I'm wondering, why the Fedora container? Can't the Ubuntu runner directly execute molecule?
- I see molecule is executed with `--debug`. Is this because the Output is really useful or is this a debugging leftover from yours? The output is very verbose currently.

thozza commented:
> Thanks a lot! This made me reading about GitHub Actions and molecule. Now I got two new questions sweat_smile
> 
> * We are three levels deep into containers here. The Ubuntu runner (which is actually a container I guess), the Fedora Docker container which runs molecule and the Podman container with the system under test. I'm wondering, why the Fedora container? Can't the Ubuntu runner directly execute molecule?

It is "just" two levels deep :sweat_smile: "Ubuntu" should be a VM per GitHub's documentation (e.g. https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources).

Although the "ubuntu-latest" is supposed to be the latest Ubuntu, it is still LTS version. Using Fedora latest container to actually run molecule has IMO benefit of having the latest Ansible and Molecule installed and used to run test cases.

I come from the Fedora land, been there a packager and contributor for over 10 years, so using Fedora is also a matter of personal preference. I know people who maintain software in Fedora, know how it is packaged, how to self-patch RPMs and I'm more closely following what is happening there. On the ohter hand, I never used Ubuntu or Debian-based OSes beyond running them on RPi and didn't want to spend more time than necessary on trying to make things work on Ubuntu.

Another reason is that I was developing the Molecule test cases on my Fedora workstation, where I use podman, so I ended up using Fedora as the container to run Molecule in GH Action to have the same environment. I tried using some of the existing GH Actions with Molecule (https://github.com/marketplace/actions/test-ansible-roles-with-molecule or https://github.com/marketplace/actions/ansible-molecule). All of them do basically the same thing - execute Molecule inside a prebuilt container.

So after spending enough time trying to use what exists, I reverted to the NIH part of myself and implemented the whole thing on my own :laughing: 

> * I see molecule is executed with `--debug`. Is this because the Output is really useful or is this a debugging leftover from yours? The output is very verbose currently.

Partially yes. I added the `--debug` option while going through all the "fun" to make the test work in GH Action, but then left it there because I thought that it may be beneficial for CI. Mainly because as long the test passes, nobody will open the job log and if it fails, the debug log may be useful. Nevertheless I would be happy to disable it :wink:

exploide commented:
> Although the "ubuntu-latest" is supposed to be the latest Ubuntu, it is still LTS version. Using Fedora latest container to actually run molecule has IMO benefit of having the latest Ansible and Molecule installed and used to run test cases.

Ok, that's fine. I think we could also use pip to install the latest Ansible and molecule versions on Ubuntu, but let's leave it as is.

> I come from the Fedora land...

Yeah, me too. In the first place I thought, why is it a Ubuntu runner at all, why not Fedora :joy: Then I learned the runners GH provides are limited and everything else needs to be a container.

Great, then I will merge this, enable GH Actions and prepare a new release. Thanks again for your work and the patience :)
